### PR TITLE
Use setup_scm, publish to PyPI and better install testing

### PR DESCRIPTION
**Pull request by mithro**
- Use `setuptool_scm` to generate the version number.
 - Add GitHub workflow ([`SymbiFlow/actions/includes/workflows/python/build-and-upload-for-pypi-pure@main`](https://github.com/SymbiFlow/actions/blob/main/includes/workflows/python/build-and-upload-for-pypi-pure/workflow.yaml)) for publishing to PyPi.
 - Add GitHub workflow ([`SymbiFlow/actions/blob/main/includes/workflows/python/install-and-test@main`](https://github.com/SymbiFlow/actions/blob/main/includes/workflows/python/install-and-test/workflow.yaml)) for installing the package in multiple ways and then running the tests.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -15,7 +43,13 @@
     python_requires=">=3.7",
     packages=setuptools.find_packages(),
     include_package_data=True,
+    setup_requires=[
+        "setuptools>=42",
+        "setuptools_scm",
+        "wheel",
+    ],
     install_requires=[
+        "importlib_metadata; python_version<'3.8'",  # for __version__
```
Comment (litghost) [inline setup.py]: Can you explain what is going on here?

**Comment by litghost** [comment]
@mithro Do you have a link to a Github actions run that is green?

**Comment by mithro** [comment]
WIP at https://github.com/mithro/symbiflow-rr-graph/actions

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/mithro/symbiflow-rr-graph">mithro/symbiflow-rr-graph</a></strong></div><div>Collection of Routing Resources Graph (RR Graph) libraries for VPR - mithro/symbiflow-rr-graph</div></blockquote>

**Comment by mithro** [comment]
Their is a related pull request at https://github.com/SymbiFlow/actions/pull/8

**Comment by mithro** [comment]
- "Test after installing from.." workflow is at https://github.com/mithro/symbiflow-rr-graph/runs/2100624397?check_suite_focus=true
 - "Build (pure python) packages and publish on PyPI" workflow is at https://github.com/mithro/symbiflow-rr-graph/actions/runs/648229214

<blockquote><div> GitHub</div><div><strong><a href="https://github.com">Build software better, together</a></strong></div><div>GitHub is where people build software. More than 56 million people use GitHub to discover, fork, and contribute to over 100 million projects.</div></blockquote>
<blockquote><div> GitHub</div><div><strong><a href="https://github.com/mithro/symbiflow-rr-graph">mithro/symbiflow-rr-graph</a></strong></div><div>Collection of Routing Resources Graph (RR Graph) libraries for VPR - mithro/symbiflow-rr-graph</div></blockquote>

**Comment by mithro** [comment]
https://github.com/SymbiFlow/symbiflow-rr-graph/pull/11 should be merged before this PR.

**Comment by mithro** [comment]
FYI - @acomodi / @mkurc-ant

**Comment by mithro** [comment]
This pull request needs to be rebased on top of #11.

**Comment by litghost** [comment]
> This pull request needs to be rebased on top of #11.

Should rebase this now?

**Review by litghost — COMMENTED** [review]
Needs to be rebased?

**Comment by mithro** [comment]
This change has now been rebased and seems to be working!

**Comment by mithro** [comment]
FYI The commits that are interesting are;
 - https://github.com/SymbiFlow/symbiflow-rr-graph/pull/5/commits/5702f1696986aa1f49a52bad1f10125c713c2baa -- Use setuptools_scm for the version number.
 - https://github.com/SymbiFlow/symbiflow-rr-graph/pull/5/commits/3ae4ab1e819df2a23bfd9675397a4d9adf2e920f -- Adding workflow to publish to PyPI.
 - https://github.com/SymbiFlow/symbiflow-rr-graph/pull/5/commits/1aeefdcbebfc737bdd1be69d1ce504537c403990 -- Adding workflow to install from multiple sources and run the test.

**Comment by mithro** [comment]
FYI - I'll probably move the [`download-and-run-tests` action](https://github.com/mithro/symbiflow-rr-graph/tree/master/.github/includes/actions/download-and-run-tests) into [SymbiFlow actions](https://github.com/SymbiFlow/actions) if it turns out to be useful for testing the other Python modules we have.

File: .github/includes/actions/download-and-run-tests/install-test-requirements.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+from __future__ import print_function
+
+import os
+import pprint
+import urllib
+import urllib.request
+import subprocess
+import sys
+
+on_ci = os.environ.get('CI', 'false')
```
Comment (litghost) [inline .github/includes/actions/download-and-run-tests/install-test-requirements.py:21]: This all oaught to be in a function (e.g. `main`).

File: .github/includes/actions/download-and-run-tests/install-test-requirements.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+# Get the requirements.txt file contents.
+if os.path.exists('requirements.txt'):
+    with open('requirements.txt') as f:
+        data = f.readlines()
+else:
+    # Download the requirements.txt file
+    assert on_ci == 'true', on_ci
```
Comment (litghost) [inline .github/includes/actions/download-and-run-tests/install-test-requirements.py:29]: Can you explain under what circumstances this else branch make sense?

File: .github/includes/actions/checkout/action.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# Copyright (C) 2021  The SymbiFlow Authors.
```
Comment (litghost) [inline .github/includes/actions/checkout/action.yaml:1]: I'm confused why this is here?  Maybe a small comment to explain that?

File: .github/includes/actions/download-and-run-tests/run-tests.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+import pprint
+import urllib
+import urllib.request
+import os.path
+import sys
+
+from pkg_resources import get_distribution
+
+# Download pytest.ini
```
Comment (litghost) [inline .github/includes/actions/download-and-run-tests/run-tests.py:22]: Please move into function

File: requirements.txt [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # Install setup_requires
 setuptools>=42
+setuptools_scm
```
Comment (litghost) [inline requirements.txt:3]: Shouldn't this also be in `pyproject.toml`?

**Review by litghost — COMMENTED** [review]
Mostly LGTM.  I've not seen all of this stuff before, so some more comments would be nice, especially if this PR is going to be the template for future PyPI integrations.

**Comment by mithro** [comment]
I started a doc @ https://docs.google.com/document/d/1wfHadeLrFsiLOsvsjQSSELzVzzNHUcPy5athVUL7Xjo/edit# include some "best practices".

<blockquote><div>Google Docs</div><div><strong><a href="https://docs.google.com/document/d/1wfHadeLrFsiLOsvsjQSSELzVzzNHUcPy5athVUL7Xjo/edit?usp=embed_facebook">SymbiFlow Python Modules Common Setup and Configuration</a></strong></div><div>SymbiFlow Python Modules Setup and Configuration  Environment and Requirements All package install requirements should be specified in the setup.py, “install_requires” # setup.pysetuptools.setup(    ...    install_requires=[        <requirements>    ],    ...)  Should have a requiremen...</div></blockquote>

**Pull request merged**